### PR TITLE
[WIP] Add example: animate a line

### PR DESCRIPTION
**Pull request by YunjieLi**
Per #1513 I created a new example of animating a line.

I rewrote the [Mpabox.js example](https://www.mapbox.com/mapbox.js/example/v1.0.0/dynamically-drawing-a-line/) using GL JS and requestAnimationFrame so it's smoother. I also add the play and stop buttons which is not in the [point animation example](https://www.mapbox.com/mapbox-gl-js/example/animate-point-along-line/).

[Here is a demo](https://bl.ocks.org/YunjieLi/5919254f2e8a517a2c7350da4598bccb).

Still working on:
[ ] Slight flickering when I stop and resume animation
[ ] Bug in leaving the tab inactive and come back
[ ] Pan the map to follow animation: speed control?

cc @geografa @mollymerp

### CODE REVIEW

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 1]
```diff
@@ -0,0 +1,128 @@
+--- layout: example category: example title: Animate a line description: Animate a line by updating a GeoJSON source on each frame. tags: - layers - sources ---
```
Comment (mollymerp) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: this section needs to be formatted like the other examples (not all on one line, etc)

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+#play.pause::after {
+  content: 'Play';
+}
+</style>
+
+<div id='map'></div>
+<button id='play' class='pause'></button>
+
+<script>
+mapboxgl.accessToken = 'pk.eyJ1IjoieXVuamllbGkiLCJhIjoiY2lxdmV5MG5rMDAxNmZta3FlNGhyMmpicSJ9.CTEQgAyZGROcpJouZuzJyA';
```
Comment (mollymerp) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: you don't need an access token here. a token is automatically added by jekyll when it compiles the examples

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    'source': 'line-animation',
+    'layout': {
+      'line-cap': 'round',
+      'line-join': 'round'
+    },
+    'paint': {
+      'line-color': '#3cc',
+      'line-width': 5,
+      'line-opacity': .4
+    }
```
Comment (mollymerp) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: This is hard to see on the map -- I would make it a brighter color or more opaque and/or change the basemap.

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    // append new coordinates to the lineStrin
+    // then update the map
+    geojson.features[0].geometry.coordinates.push([x, y]);
+    map.getSource('line-animation').setData(geojson);
+
+    // Fly the map to follow the animation
+    // map.panBy([60, 0]);
+
+    // Request the next frame of the animation.
+    animation = requestAnimationFrame(animateLine);
```
Comment (lucaswoj) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: There's some interesting discussion about using using `requestAnimationFrame` with `GeoJSONSource#setData` in https://github.com/mapbox/mapbox-gl-js/issues/3692

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    // to cancel the animation:
+    if (this.classList.contains('pause')) {
+      cancelAnimationFrame(animation);
+      // to play the animation
+    } else {
+      // resume previous progress if progress != 0 
+      startTime = performance.now() - progress;
+      animation = requestAnimationFrame(animateLine);
+    }
+  });
```
Comment (lucaswoj) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: Would you be open to removing the "play" button in order to reduce the # of lines of code in this example?

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        }]
+      };
+    }
+
+    progress = time - startTime;
+    // speedFactor controls the animation speed.
+    var x = progress / speedFactor;
+    // draw a sine wave with some math.
+    var y = Math.sin(x * Math.PI / 90) * 40;
+    // append new coordinates to the lineStrin
```
Comment (lucaswoj) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: You're missing a "g" on "line string"

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            "type": "LineString",
+            "coordinates": [
+              [0, 0]
+            ]
+          }
+        }]
+      };
+    }
+
+    progress = time - startTime;
```
Comment (kronick) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: This is the source of the inactive tab bug-- when on another tab, requestAnimationFrame does not get called, so your progress variable makes a big jump when you return and you don't add any points in between. Consider something like this to limit the maximum timestep between frames:

    var startTime;
    var lastTime;
    var progress = 0;

    function animateLine(time) {
      var dT = time - lastTime;
      if(dT > 30) dT = 30; // Don't jump too far between animation frames, even if the browser tab is inactive
      progress += dT;
      lastTime = time;
   
      // The rest of what you have should work the same

    }

File: docs/_posts/examples/3400-01-31-animate-a-line.html (hunk 2 above)
Comment (YunjieLi) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: Good catch! I forgot to remove it from the gist

File: docs/_posts/examples/3400-01-31-animate-a-line.html (hunk 5 above)
Comment (YunjieLi) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: I added it according to @geografa's [comment](https://github.com/mapbox/mapbox-gl-js/issues/1513#issuecomment-266136120) that user wants more control over the animation. But I do think I can simplify into just a 'play/replay' button + no looping for animations. Do you think that would make it clean enough?

File: docs/_posts/examples/3400-01-31-animate-a-line.html (hunk 7 above)
Comment (YunjieLi) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: Thanks @kronick @tmcw! I'll research into both suggestions a little more. I think `visibilitychange` sounds more straightforward but am not sure if there are other things that can interrupt `requestAnimationFrame` besides inactive tabs.

File: docs/_posts/examples/3400-01-31-animate-a-line.html [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    };
+    // play animation
+    animateLine();
+  });
+
+  function animateLine(timestamp) {
+
+    // requestAnimationFrame pauses when the tab is inactive
+    // reset startTime and progress when switched back
+    document.addEventListener('visibilitychange', function(){
```
Comment (tmcw) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: this call shouldn't be inside of `animateLine`: `animateLine` will be called many times but `document.addEventListener` should be called once.

Comment (YunjieLi) [inline docs/_posts/examples/3400-01-31-animate-a-line.html]: Thanks @tmcw !  I just pushed some new edits if you want to take a look again?

**Comment by YunjieLi** [comment]
@mollymerp which branch should I rebase this to?

**Comment by mollymerp** [comment]
@YunjieLi you can rebase against `mb-pages` 👍

**Comment by YunjieLi** [comment]
Also - currently the map shows a straight line between the starting point and x,y by the time the window is loaded, versus not starting animation until the map is ready(?). Will look into it next.
image

**Comment by YunjieLi** [comment]
Somehow I couldn't get this running locally anymore, so moving to https://github.com/mapbox/mapbox-gl-js/pull/4780!

**Pull request closed without merging**